### PR TITLE
Don't create workers ASG if the cluster version is cloud formation

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#2023

This PR adds a UseCloudFormation key helper we'll use to enable / disable logic while we migrate to CF. It returns true when the cluster version is set to "cloud-formation" which we can do via the helm local chart.

The workers ASG is disabled in the legacy resource if the CF version is set. We will expand this as more resources are migrated to the CF templates.

@xh3b4sd This is just an FYI on the approach.

### CODE REVIEW

File: service/cloudconfig/cloud_config.go [hunk 1]
```diff
@@ -56,8 +56,9 @@ func New(config Config) (*CloudConfig, error) {
 func (c *CloudConfig) NewMasterTemplate(customObject awstpr.CustomObject, certs certificatetpr.CompactTLSAssets, keys randomkeytpr.CompactRandomKeyAssets) (string, error) {
 	var err error
 
-	// TODO remove defaulting as soon as custom objects are configured.
-	if !key.HasClusterVersion(customObject) {
+	// Default the version if it is not configured or we are using Cloud Formation.
+	// TODO Remove once Cloud Formation migration is complete.
```
Comment (rossf7) [inline service/cloudconfig/cloud_config.go]: We're using the version to trigger the Cloud Formation logic. So we need to set the version here to get the correct cloud config.

File: service/key/key.go [hunk 2]
```diff
@@ -8,6 +8,13 @@ import (
 	"github.com/giantswarm/microerror"
 )
 
+const (
+	// cloudFormationVersion is set during the migration so resources are
+	// managed by the cloudformation resource.
+	// TODO Remove once the migration is complete.
+	cloudFormationVersion = "cloud-formation"
```
Comment (rossf7) [inline service/key/key.go]: Setting this version in the custom object triggers the logic.

File: service/resource/legacy/resource.go [hunk 3]
```diff
@@ -821,91 +821,94 @@ func (s *Resource) processCluster(cluster awstpr.CustomObject) error {
...
-		lcInput.associatePublicIP = true
-		lcInput.subnet = publicSubnet
-	}
-
-	// An EC2 Keypair is needed for legacy clusters. New clusters provide SSH keys via cloud config.
-	if !key.HasClusterVersion(cluster) {
-		lcInput.keypairName = key.ClusterID(cluster)
-	}
+	// During Cloud Formation migration this logic is only needed for non CF clusters.
+	if !key.UseCloudFormation(cluster) {
```
Comment (rossf7) [inline service/resource/legacy/resource.go]: The workers launch config and ASG logic is disabled and provided by the CF resource. We will expand this as we migrate more resources to the CF templates.

File: service/resource/legacy/resource.go [hunk 4]
```diff
@@ -991,28 +994,31 @@ func (s *Resource) processDelete(cluster awstpr.CustomObject) error {
 		s.logger.Log("info", "deleted masters")
 	}
 
-	// Delete workers Auto Scaling Group.
-	asg := awsresources.AutoScalingGroup{
-		Client: clients.AutoScaling,
-		Name:   key.AutoScalingGroupName(cluster, prefixWorker),
-	}
+	// During Cloud Formation migration this logic is only needed for non CF clusters.
+	if !key.UseCloudFormation(cluster) {
```
Comment (rossf7) [inline service/resource/legacy/resource.go]: Same applies for deleting the workers ASG and launch config.

**Comment by xh3b4sd** [comment]
My only take on this is that it is ok to use these hacks for the transitioning but once we version resources via version bundles we will stop having these hacky versions exceptions within the code. Later there will only be a versioned resource that implements business logic without such hacks and exceptions.

**Comment by rossf7** [comment]
Yes 💯  this is just to help with the transitioning. The cloudformation resource should use version bundles and we can get rid of this when we remove the legacy resource.

**Pull request merged**